okhttp-clients/src/main/java/com/palantir/conjure/java/okhttp/UrlSelectorImpl.java

```diff
@@ -28,19 +29,25 @@
 import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
+import java.util.concurrent.ThreadLocalRandom;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
 import okhttp3.HttpUrl;
 
 final class UrlSelectorImpl implements UrlSelector {
-
-    private final ImmutableList<HttpUrl> baseUrls;
+    private final Supplier<List<HttpUrl>> baseUrls;
     private final AtomicInteger currentUrl;
     private final Cache<HttpUrl, UrlAvailability> failedUrls;
     private final boolean useFailedUrlCache;
 
-    private UrlSelectorImpl(ImmutableList<HttpUrl> baseUrls, Duration failedUrlCooldown) {
-        this.baseUrls = baseUrls;
+    private UrlSelectorImpl(ImmutableList<HttpUrl> baseUrls, boolean randomizeOrder, Duration failedUrlCooldown) {
+        this.baseUrls = Suppliers.memoizeWithExpiration(
+                () -> randomize(baseUrls, randomizeOrder),
+                // Add jitter to avoid mass node reassignment when multiple nodes of a client are restarted
+                600 + ThreadLocalRandom.current().nextLong(-30, 30),
```

Comment:
I think we should justify the choice of 10 minutes here (especially as our read timeouts are in the same order of magnitude).  It seems to me that we just want clients to randomize more frequently than services are restarted... should we model this as roughly daily?

In the case of a two-node service, after one node switches off, the other node can be expected to receive 100% of traffic from all clients until half of this duration has elapsed.  Given that many our services already seem to handle the degenerate case of 99% load on one node and 1% on the other, maybe we could actually be more conservative with this?

What do you think about using something like 6 hours instead of 10 minutes?